---
title: Promise
date: 2023-10-03 17:19:51
permalink: /pages/0033c7/
categories:
  - 前端基础
  - Javascript
tags:
  - 
author: 
  name: cyy
  link: https://github.com/Bunny0927
---

## promise出现的原因
异步处理事件和回调的问题：
- **嵌套调用/回调地狱**：下一个任务依赖上一个任务的请求结果，并在上一个任务的回调函数内部执行新的业务逻辑，当嵌套层次多了之后，代码的可读性就变得非常差；
- **任务的不确定性**：执行每个任务都有两种可能的结果（成功或者失败），所以对每个任务都要进行一次额外的错误处理的方式，其明显增加了代码的混乱程度；

ES6引入promise，为了解决以下两个问题：
- **减少嵌套调用**
- **合并多个任务的错误处理**

## promise基础
promise是异步编程的一种解决方案。

### 两个参数：resolve、reject；三种状态：pending、fulfilled（resolved）、rejected。
promise构造函数接受一个函数作为参数，该函数的的两个参数分别为resolve、reject。它们是两个函数。

如何改变状态：
1. `resolve(value)`：如果当前状态是`pending`就会变为`resolved`
2. `reject(reason)`：如果当前状态是`pending`就会变为`rejected`
3. 抛出异常：如果当前是`pending`就会变为`rejected`

```js
// 创造promise实例
const promise = new Promise((resolve, reject) => {
  // ...
  if (/* 异步操作成功 */) {
    resolve(value)
  } else {
    reject(error)
  }
})
```
1. Promise.resolve
  将现有对象转为 Promise 对象。
  ```js
  Promise.resolve('foo')
  // 等价于
  new Promise(resolve => resolve('foo'))
  ```
  Promise.resolve方法的参数分成四种情况。
  - 参数是一个 Promise 实例 => 原封不动地返回这个实例。
  - 参数是一个thenable对象 => Promise.resolve方法会将这个对象转为 Promise 对象，然后就立即执行thenable对象的then方法。
  - 参数不是具有then方法的对象，或根本就不是对象 => Promise.resolve方法返回一个新的 Promise 对象，状态为resolved。
  - 不带有任何参数 => Promise.resolve方法允许调用时不带参数，直接返回一个resolved状态的 Promise 对象。
2. Promise.reject
   返回一个新的实例，状态为rejected。
   ```js
    const p = Promise.reject('出错了');
    // 等同于
    const p = new Promise((resolve, reject) => reject('出错了'))

    p.then(null, function (s) {
      console.log(s)
    });
    // 出错了
   ```
   Promise.reject()方法的参数，会原封不动地作为reject的理由，变成后续方法的参数。
  
### 基础功能：then、catch、finally

1. then()
   为 Promise 实例添加状态改变时的回调函数，then方法的第一个参数是resolved状态的回调函数，第二个参数是rejected状态的回调函数，它们都是可选的。
   ```js
    promise.then(function(value) {
      // success
    }, function(error) {
      // failure
    })
    ```
    then方法返回的是一个新的Promise实例（注意，不是原来那个Promise实例）。因此可以采用链式写法，即then方法后面再调用另一个then方法。
    ```js
   ```
2. catch()
   用于指定发生错误时的回调函数
   一般来说，不要在then()方法里面定义 Reject 状态的回调函数（即then的第二个参数），总是使用catch方法。
    ```js
    promise.then(function(value) {
      // success
    }).catch(function(error) {
      // failure
    })
    ```
    catch方法返回的还是一个 Promise 对象，因此后面还可以接着调用then方法。
    ```js
3. finally()
   不管 Promise 对象最后状态如何，都会执行的操作
    ```js
    promise.then(function(value) {
      // success
    }).catch(function(error) {
      // failure
    }).finally(function() {
      // 无论成功还是失败都会执行
    })
    ```

### 特殊方法：all()、race()、allSettled()、any()
用于将多个 Promise 实例，包装成一个新的 Promise 实例。
1. all()
   p的状态由p1、p2、p3决定，分成两种情况。
   - 只有p1、p2、p3的状态都变成fulfilled，p的状态才会变成fulfilled，此时p1、p2、p3的返回值组成一个数组，传递给p的回调函数。
   - 只要p1、p2、p3之中有一个被rejected，p的状态就变成rejected，此时第一个被reject的实例的返回值，会传递给p的回调函数。   
   ```js
   const p = Promise.all([p1, p2, p3])
   p.then(function (value) {
     // ...
   }, function (reason) {
     // ...
   })
   ```
2. race()
    只要p1、p2、p3之中有一个实例率先改变状态，p的状态就跟着改变。那个率先改变的 Promise 实例的返回值，就传递给p的回调函数。

    ```js
    const p = Promise.race([p1, p2, p3]);
    ```

    例子：如果 5 秒之内fetch方法无法返回结果，变量p的状态就会变为rejected，从而触发catch方法指定的回调函数。
    ```js
    const p = Promise.race([
      fetch('/resource-that-may-take-a-while'),
      new Promise(function (resolve, reject) {
        setTimeout(() => reject(new Error('request timeout')), 5000)
      })
    ]);
    p
    .then(console.log)
    .catch(console.error);
    ```
3. allSettled()
   用来确定一组异步操作是否都结束了（不管成功或失败）。所以，它的名字叫做“Settled”，包含了“fulfilled”和“rejected”两种情况。

   Promise.allSettled()方法接受一个数组作为参数，数组的每个成员都是一个 Promise 对象，并返回一个新的 Promise 对象。只有等到参数数组的所有 Promise 对象都发生状态变更（不管是fulfilled还是rejected），返回的 Promise 对象才会发生状态变更。

    ```js
    const resolved = Promise.resolve(42);
    const rejected = Promise.reject(-1);

    const allSettledPromise = Promise.allSettled([resolved, rejected]);

    allSettledPromise.then(function (results) {
      console.log(results);
    });
    // [
    //    { status: 'fulfilled', value: 42 },
    //    { status: 'rejected', reason: -1 }
    // ]
    ```
4. any()
   只要参数实例有一个变成fulfilled状态，包装实例就会变成fulfilled状态；如果所有参数实例都变成rejected状态，包装实例就会变成rejected状态。
   
   Promise.any()跟Promise.race()方法很像，只有一点不同，就是Promise.any()不会因为某个 Promise 变成rejected状态而结束，必须等到所有参数 Promise 变成rejected状态才会结束。
   
   例子：Promise.any()方法的参数数组包含三个 Promise 操作。其中只要有一个变成fulfilled，Promise.any()返回的 Promise 对象就变成fulfilled。如果所有三个操作都变成rejected，那么await命令就会抛出错误。
   ```js
    Promise.any([
      fetch('https://v8.dev/').then(() => 'home'),
      fetch('https://v8.dev/blog').then(() => 'blog'),
      fetch('https://v8.dev/docs').then(() => 'docs')
    ]).then((first) => {  // 只要有一个 fetch() 请求成功
      console.log(first);
    }).catch((error) => { // 所有三个 fetch() 全部请求失败
      console.log(error);
    });
   ```

## promise使用
1. 首先初始化一个 Promise 对象，可以通过两种方式创建，这两种方式都会返回一个 Promise 对象。
   - new Promise(fn)
   - Promise.resolve(fn)

2. 然后调用上一步返回的 promise 对象的 then 方法，注册回调函数。
   then 中的回调函数可以有一个参数，也可以不带参数。如果 then 中的回调函数依赖上一步的返回结果，那么要带上参数。比如
   ```js
   new Promise(fn)
    .then(fn1(value）{
        //处理value
    })
   ```  

3. 最后注册 catch 异常处理函数，处理前面回调中可能抛出的异常。

## promise与事件循环
Promise在初始化时，传入的函数是同步执行的，然后注册 then 回调。
注册完之后，继续往下执行同步代码，在这之前，then 中回调不会执行。
同步代码块执行完毕后，才会在事件循环中检测是否有可用的 promise 回调。
如果有，那么执行，如果没有，继续下一个事件循环。

## promise与async/await
async/await是promise的语法糖，是基于promise实现的。
async/await是一种更加优雅的异步编程方式，可以让我们用同步的方式去写异步代码。
```js
async function async1() {
  console.log('async1 start')
  await async2()
  console.log('async1 end')
}

// async/await 异步方法同步执行
(async ()=>{
  await a();
  await b();
  await c();
})()
```


> 其他博客链接参考
> 
> [Promise基础以及使用示例](https://juejin.cn/post/7256594892721684541)
> 
> [阮一峰-promise对象](https://es6.ruanyifeng.com/#docs/promise)